src/OrchardCore/OrchardCore.Abstractions/Shell/Builders/Extensions/ServiceDescriptorExtensions.cs

```diff
@@ -13,19 +13,35 @@ public static Type GetImplementationType(this ServiceDescriptor descriptor)
                 return cloned.Parent.GetImplementationType();
             }
 
-            if (descriptor.ImplementationType != null)
+            if (descriptor.ServiceKey == null)
             {
-                return descriptor.ImplementationType;
+                if (descriptor.ImplementationType != null)
+                {
+                    return descriptor.ImplementationType;
+                }
+                else if (descriptor.ImplementationInstance != null)
```

Comment:
@hishamco i can test it out tomorrow if no one else does and merge meanwhile. If @jtkech approves it, you can merge it. But, here you can safely drop all `else` from the condition since all else if return

Reply:
I just updated the PR, I might need to test it against the Azure email PR